src/librustc_mir/transform/mod.rs

```diff
@@ -255,6 +255,11 @@ fn optimized_mir<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId) -> &'tcx
 
         // Optimizations begin.
         inline::Inline,
+
+        // Lowering generator control-flow and variables
+        // has to happen before we do anything else to them.
+        generator::StateTransform,
+
```

Comment:
We want to lower generator as late as possible, in particular we want to run at least `SimplifyLocals` before the `StateTransform`. Did you move this up because you triggered the sanity checks in `StateTransform`?

Reply:
Yeah, those sanity checks don't survive past optimizations.